refactor: Return `util::Result` from WalletLoader methods

w0xlt opened the pull request:
This PR adds a method that implement common logic to WalletLoader methods and change them to return `BResult<std::unique_ptr<Wallet>>`.

Motivation: #25594 changed `restoreWallet` to return `BResult` but this method shares a common pattern with  `createWallet` and `loadWallet`. This PR keeps the same pattern to all WalletLoader methods.

furszy commented on src/qt/walletcontroller.cpp:
```diff
@@ -343,9 +345,11 @@ void OpenWalletActivity::open(const std::string& path)
     QTimer::singleShot(0, worker(), [this, path] {
-        std::unique_ptr<interfaces::Wallet> wallet = node().walletLoader().loadWallet(path, m_error_message, m_warning_message);
+        auto wallet{node().walletLoader().loadWallet(path, m_warning_message)};
 
-        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(wallet));
+        m_error_message = wallet ? bilingual_str{} : wallet.GetError();
+
+        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(wallet.ReleaseObj());
```
```suggestion
auto wallet{node().walletLoader().loadWallet(path, m_warning_message)};

if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(wallet.ReleaseObj());
else m_error_message = wallet.GetError();

w0xlt replied:
Done in https://github.com/bitcoin/bitcoin/pull/25616/commits/948651da4f88b3479f36f48163f09bf2af4d2c92.

furszy commented on src/qt/walletcontroller.cpp:
```diff
@@ -262,9 +262,11 @@ void CreateWalletActivity::createWallet()
     QTimer::singleShot(500ms, worker(), [this, name, flags] {
-        std::unique_ptr<interfaces::Wallet> wallet = node().walletLoader().createWallet(name, m_passphrase, flags, m_error_message, m_warning_message);
+        auto wallet{node().walletLoader().createWallet(name, m_passphrase, flags, m_warning_message)};
 
-        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(wallet));
+        m_error_message = wallet ? bilingual_str{} : wallet.GetError();
+
+        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(wallet.ReleaseObj());
```
```suggestion
auto res_wallet{node().walletLoader().createWallet(name, m_passphrase, flags, m_warning_message)};

if (res_wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(res_wallet.ReleaseObj());
else m_error_message = res_wallet.GetError();
```

furszy reviewed (commented):
left a quick review

w0xlt commented:
Thanks  @furszy for the review. The suggestions were addressed in https://github.com/bitcoin/bitcoin/commit/948651da4f88b3479f36f48163f09bf2af4d2c92 and https://github.com/bitcoin/bitcoin/commit/3c96b258e530bf0f99a4a0c7ddc2d5dbca732119.

MarcoFalke commented on src/qt/walletcontroller.cpp:
```diff
@@ -343,9 +344,10 @@ void OpenWalletActivity::open(const std::string& path)
 
-        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(wallet));
+        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(wallet.ReleaseObj());
+        else m_error_message = wallet.GetError();
```
Below I used `        m_error_message = wallet ? bilingual_str{} : wallet.GetError();`

It would be good to use the same code consistently everywhere.

shaavan replied:
I think the code change was not correctly pushed.
The commit 7cbfa1335bd89d5831e916672e57aa8601b12bb4 still displays the line:

```
else m_error_message = wallet.GetError();
```
and not:
```
m_error_message = wallet ? bilingual_str{} : wallet.GetError();
```

w0xlt replied:
@shaavan I said the suggestion was addressed because I understand it's about keeping all code consistent across all commits.

I had originally defined all  `m_error_message` same way as done in https://github.com/bitcoin/bitcoin/pull/25594: ` m_error_message = wallet ? bilingual_str{} : wallet.GetError();`.

So https://github.com/bitcoin/bitcoin/pull/25616#discussion_r921692400 and https://github.com/bitcoin/bitcoin/pull/25616#discussion_r921688995 suggested changing them to `else m_error_message = res_wallet.GetError();` but I had forgotten to change it to `restoreWallet()`. https://github.com/bitcoin/bitcoin/commit/7cbfa1335bd89d5831e916672e57aa8601b12bb4 fixed this.

If I understand the `BResult` interface correctly, it seems to be the clearest option, as [`BResult<T>::GetError()`](https://github.com/bitcoin/bitcoin/blob/826fae6a0fe184518af625ebb6d1ef9e64631385/src/util/result.h#L41-L44) will return the default `bilingual_str{} :` value [if `T` is not set](https://github.com/bitcoin/bitcoin/blob/826fae6a0fe184518af625ebb6d1ef9e64631385/src/util/result.h#L22).

However, both approaches work. If the reviewers have a strong opinion about one of them, I can change it.

furszy replied:
`BResult` is essentially an std::variant wrapper. It stores one object OR the other (succeed or failure). Not both.

So, it is redundant to ask `m_error_message = wallet ? bilingual_str{} : wallet.GetError();.` first and then ask if succeeded in another if block.

MarcoFalke replied:
`m_error_message = wallet ? bilingual_str{} : wallet.GetError();` was intentional to reset the member that stores the error, instead of leaving the previous value, if no error occurred. However, I am fine changing it to something else.

furszy replied:
Yeah, I double checked it before comment. The error message is only set here after calling the backend method. So, the string will always be empty if no error occurs.

MarcoFalke replied:
Is it possible to load a wallet with an error and then load a wallet without error on the same walletcontroller?

furszy replied:
Oh, I missed the comment, sorry Marko.

------

> Is it possible to load a wallet with an error and then load a wallet without error on the same walletcontroller?

You mean, restoring or opening a wallet?

Because if that is the case, there shouldn't be any problem. Every time that we restore or open a wallet we create a new `WalletControllerActivity` subclass which encapsulates the error and warning messages (it's a single shot class. It gets deleted as soon as it finishes processing the action).

w0xlt commented:
https://github.com/bitcoin/bitcoin/pull/25616#discussion_r921916024 addressed in https://github.com/bitcoin/bitcoin/commit/7cbfa1335bd89d5831e916672e57aa8601b12bb4.

aureleoules commented on src/wallet/interfaces.cpp:
```diff
@@ -549,8 +549,15 @@ class WalletLoaderImpl : public WalletLoader
     void setMockTime(int64_t time) override { return SetMockTime(time); }
 
+    //! Return a wallet interface object or an error wrapped in a BResult instance
+    BResult<std::unique_ptr<Wallet>> makeWallet(const std::shared_ptr<CWallet>& _wallet, bilingual_str& error)
```
can `makeWallet` be renamed to `makeWalletResult` or something else? Having both `createWallet` and `makeWallet` is a bit confusing.

w0xlt replied:
Done in https://github.com/bitcoin/bitcoin/pull/25616/commits/087959d6e92ed37b72b86dbbc9b0ff052a6c5a75. Thanks.

aureleoules reviewed (commented):
ACK 7cbfa1335bd89d5831e916672e57aa8601b12bb4.
I think this is a good use-case of BResult.

nit: Shouldn't the commits be merged into one as the changes are small and related?

shaavan reviewed (commented):
Concept ACK 

- I agree with the idea of using **BResult** as the return type for walletloader method, as it ensures that the return value has either the result value (if succeed) or the error value (if failure) and never both.
- The code changes look clean and concise. I want to look further into [this](https://github.com/bitcoin/bitcoin/pull/25616#discussion_r923353482) comment before ACKing the correctness of code change.
> Is it possible to load a wallet with an error and then load a wallet without an error on the same walletcontroller?

- In the meantime, I would suggest squashing the commits together.

w0xlt commented:
https://github.com/bitcoin/bitcoin/pull/25616#pullrequestreview-1041617804 and https://github.com/bitcoin/bitcoin/pull/25616#pullrequestreview-1041856598 were addressed in https://github.com/bitcoin/bitcoin/commit/087959d6e92ed37b72b86dbbc9b0ff052a6c5a75.

shaavan reviewed (approved):
ACK 087959d6e92ed37b72b86dbbc9b0ff052a6c5a75

Changes since my last review:
- Renamed `makeWallet` -> `makeWalletResult`.
- Squashed commits.

As per @furszy explanation in [this](https://github.com/bitcoin/bitcoin/pull/25616#discussion_r928066324) comment, since the `WalletControllerActivity` class gets deleted after the loading of a wallet finishes, it is not possible to have an error message stored from a previous erroneous loading of a wallet. 

Hence it is not necessary to manually reset `bilingual_str{}`.

Thanks, @furszy, for the explanation.

jonatack commented on src/wallet/interfaces.cpp:
```diff
@@ -549,8 +549,15 @@ class WalletLoaderImpl : public WalletLoader
+    {
+        BResult<std::unique_ptr<Wallet>> wallet{MakeWallet(m_context, _wallet)};
+        if (!wallet) return error;
+        return wallet;
```
nit, simpler with one `return`, feel free to ignore
```diff
-        if (!wallet) return error;
-        return wallet;
+        return wallet ? wallet : error;
```

jonatack reviewed (commented):
Approach ACK

w0xlt committed:
wallet: Return `util::Result` from WalletLoader methods

w0xlt commented:
Thanks for the review @shaavan, but a rebase was needed.

Rebased and changed the code to use `util::Result` which replaced the `BResult`.

@jonatack I used the suggested format in this new commit.

jonatack commented on src/interfaces/wallet.h:
```diff
@@ -88,7 +88,7 @@ class Wallet
 
     // Get a new address.
-    virtual util::Result<CTxDestination> getNewDestination(const OutputType type, const std::string label) = 0;
+    virtual util::Result<CTxDestination> getNewDestination(const OutputType type, const std::string& label) = 0;
```
Thanks for picking up https://github.com/bitcoin/bitcoin/pull/25721#discussion_r938747475 here and fixing up indentation in this file.

MarcoFalke replied:
The wallet method is still creates a copy?

```
$ git grep GetNewDestination -- '*wallet.h'
src/wallet/wallet.h:    util::Result<CTxDestination> GetNewDestination(const OutputType type, const std::string label);

jonatack commented on src/wallet/interfaces.cpp:
```diff
@@ -551,31 +551,42 @@ class WalletLoaderImpl : public WalletLoader
         options.require_existing = true;
-        return MakeWallet(m_context, LoadWallet(m_context, name, true /* load_on_start */, options, status, error, warnings));
+        bilingual_str error;
+        util::Result<std::unique_ptr<Wallet>> wallet{MakeWallet(m_context, LoadWallet(m_context, name, true /* load_on_start */, options, status, error, warnings))};
```
nit if you repush, here and in `createWallet()` above, while touching this line can use Clang-tidy named arg format (`/*load_on_start=*/`true) like in `restoreWallet()` below.

jonatack commented on src/wallet/interfaces.cpp:
```diff
@@ -551,31 +551,42 @@ class WalletLoaderImpl : public WalletLoader
+        if (!wallet) {
+            return util::Error{error};
+        }
+        return wallet;
```
nit, could save 9 lines by replacing the above 4 lines in each of the 3 methods with identical logic in one line: `return wallet ? wallet : util::Error{error};` (feel free to ignore)

w0xlt replied:
That would be good, but applying this change will result in the error below.
I'm not sure about the reason for the error.

```
wallet/interfaces.cpp:577:25: error: call to implicitly-deleted copy constructor of 'util::Result<std::unique_ptr<Wallet>>'
        return wallet ? wallet : util::Error{error};
                        ^~~~~~
./util/result.h:38:36: note: copy constructor of 'Result<std::unique_ptr<interfaces::Wallet>>' is implicitly deleted because field 'm_variant' has a deleted copy constructor
    std::variant<bilingual_str, T> m_variant;
```

jonatack replied:
Ah, `std::unique_ptr` is move constructible and move assignable but not copy constructible or copy assignable. So this would work: `return wallet ? std::move(wallet) : util::Error{error};`

w0xlt replied:
Thanks. Done in https://github.com/bitcoin/bitcoin/pull/25616/commits/466374f6c1a032d7540b1c9307631f56fe555fcf.

So `return wallet` moves the object and `return wallet ? wallet : util::Error{error};` copies the object. Interesting.

jonatack replied:
It might be that return with a ternary doesn't benefit from RVO (return value optimization) by the compiler in the same way as with an if statement.

https://en.cppreference.com/w/cpp/language/copy_elision

w0xlt replied:
Interesting discussion about this:

https://stackoverflow.com/questions/48250917/should-we-write-stdmove-in-the-cases-when-rvo-can-not-be-done

https://stackoverflow.com/questions/22078029/why-does-the-ternary-operator-prevent-return-value-optimization

jonatack replied:
Thanks, not a blocker but this is a topic I plan to go deeper into in order to improve my understanding.

jonatack reviewed (commented):
Review/debug build/unit tests ACK f179998f8072ef5f754d42b5d3d7c14e2fdfe7a3

theStack reviewed (commented):
Concept ACK

nit: Same as done in the commit message, the PR title and description should also be adapted to the latest rebase (s/`BResult`/`util::Result`/).

w0xlt commented:
CI error seems unrelated.
`interface_usdt_validation.py` worked fine on my machine.

https://cirrus-ci.com/task/6722929781112832

jonatack commented on .vscode/settings.json:
```diff
@@ -0,0 +1,81 @@
+{
```
Looks like this file invited itself into the last push :D

jonatack commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -919,10 +919,9 @@ BOOST_FIXTURE_TEST_CASE(wallet_sync_tx_invalid_state_test, TestingSetup)
 
     CMutableTransaction mtx;
-    mtx.vout.push_back({COIN, GetScriptForDestination(dest)});
+    mtx.vout.push_back({COIN, GetScriptForDestination(*op_dest)});
```
Thanks for picking up https://github.com/bitcoin/bitcoin/pull/25721#discussion_r938759439 here.

jonatack commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1770,7 +1770,7 @@ bool DescriptorScriptPubKeyMan::GetReservedDestination(const OutputType type, bo
         error = util::ErrorString(op_dest);
     }
-    return bool(op_dest);
+    return op_dest.has_value();
```
Thanks for picking up the suggestion https://github.com/bitcoin/bitcoin/pull/25721#discussion_r938825139 here.

jonatack reviewed (commented):
ACK be13477a82e4dc33a88d49dcf3f7a296bcb6c05c

furszy commented on src/qt/walletcontroller.cpp:
```diff
@@ -262,9 +262,12 @@ void CreateWalletActivity::createWallet()
+        auto wallet{node().walletLoader().createWallet(name, m_passphrase, flags, m_warning_message)};
 
-        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(wallet));
+        m_error_message = util::ErrorString(wallet);
+        if (wallet) {
+            m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(*wallet));
+        }
```
why the revert here? (and in the others as well).
the result could be an object or an error, not both.

```c++
if (wallet) m_wallet_model = something;
else m_error_message = util::ErrorString(wallet);
```

Thinking that might be good to add an assertion inside `util::ErrorString`, callers should always check that the result is an error prior to convert it.

w0xlt replied:
Maybe `assert(std::holds_alternative<bilingual_str>(result.m_variant));` inside `util::ErrorString` ?

However this assertion cause a compile-time error `error: static_assert failed due to requirement '__detail::__variant::__exactly_once<bilingual_str, bilingual_str, bilingual_str>' "T must occur exactly once in alternatives"`

furszy replied:
yep, that was where I was pointing to.
`assert(!result.has_value());` there should work fine.

w0xlt replied:
I applied the suggestion above (`if (wallet) m_wallet_model = something;`) in https://github.com/bitcoin/bitcoin/pull/25616/commits/ac7479953bc95baf6ef4b35fa25c27dd4100fe6a.

But regarding the assertion, I think a follow-up PR is better  because it also requires changing the `src/test/result_tests.cpp` which expects the result of `util::ErrorString` be `{}` if the object has value This would increase the scope of this PR.

jonatack replied:
Missing brackets in the conditional again (e.g. https://github.com/bitcoin/bitcoin/pull/25616#discussion_r938595550)

furszy replied:
shouldn't be needed:
https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?plain=1#L77-L80

jonatack replied:
"If an if only has a single-statement then-clause, it can appear on the same line as the if, without braces. In every other case, braces are required, and the then and else clauses must appear correctly indented on a new line."

furszy replied:
yep, aren't we having a single line statement here?

```c++
if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(*wallet));
else m_error_message = util::ErrorString(wallet);
```

furszy reviewed (commented):
utACK be13477a, just a small comment.

furszy reviewed (approved):
ACK ac74799

w0xlt commented:
Rebased.

jonatack commented:
Review ACK 07df6cda1468ed45ac227ac6f0169b040e5c0bf3

theStack reviewed (approved):
Code-review ACK 07df6cda1468ed45ac227ac6f0169b040e5c0bf3

MarcoFalke commented on src/interfaces/wallet.h:
```diff
@@ -320,31 +320,31 @@ class WalletLoader : public ChainClient
 public:
     //! Create new wallet.
-    virtual std::unique_ptr<Wallet> createWallet(const std::string& name, const SecureString& passphrase, uint64_t wallet_creation_flags, bilingual_str& error, std::vector<bilingual_str>& warnings) = 0;
+    virtual util::Result<std::unique_ptr<Wallet>> createWallet(const std::string& name, const SecureString& passphrase, uint64_t wallet_creation_flags, std::vector<bilingual_str>& warnings) = 0;
```
Would probably be best to also wrap the warnings in the result? Are you working on this @ryanofsky ?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25616#discussion_r942704207

> Would probably be best to also wrap the warnings in the result? Are you working on this @ryanofsky ?

It needs rebase, but yes I did this in #25722

MarcoFalke replied:
Yeah, I meant a minimal extract without the other C++ bloat (for example `Result<void>`) that isn't needed for simply passing warnings

MarcoFalke commented on src/qt/walletcontroller.cpp:
```diff
@@ -393,8 +401,11 @@ void RestoreWalletActivity::restore(const fs::path& backup_file, const std::stri
         auto wallet{node().walletLoader().restoreWallet(backup_file, wallet_name, m_warning_message)};
 
-        m_error_message = util::ErrorString(wallet);
-        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(*wallet));
```
nit: I preferred the previous code, as it is shorter and less ambiguous
